PARQUET-1678: [C++] Provide classes for reading/writing using input/output operators

gawain-bolton opened the pull request:
This commit adds two classes to the parquet namespace:
- StreamReader
- StreamWriter

These classes can be used to read/write values to Parquet data files
using a stream based approach using C++ input/output operators.

The goal is to make it easy to read/write data for applications which
deal with streams of data i.e. they do not have all data available in
order to create each complete column of data.

Checks are done at runtime to ensure that the types of variables
supplied when reading/writing match the schema.  Any difference
detected will result in a ParquetException being thrown.

Please see the
examples/parquet/parquet-stream-api/stream-reader-writer.cc file to
see how these classes can be used.

emkornfield commented on cpp/examples/parquet/CMakeLists.txt:
```diff
@@ -18,6 +18,7 @@
 add_executable(parquet-low-level-example low-level-api/reader-writer.cc)
 add_executable(parquet-low-level-example2 low-level-api/reader-writer2.cc)
 add_executable(parquet-arrow-example parquet-arrow/reader-writer.cc)
+add_executable(parquet-stream-api-example parquet-stream-api/stream-reader-writer.cc)
```
this is potentially useful, but can you also please provide proper unit tests for this PR

gawain-bolton replied:
I added unit tests for the StreamWriter class.
For the StreamReader class I have not figured out where the Parquet test data files are located.  Can you help with this?

emkornfield replied:
I think since this is a wrapper, it is probably okay to generate test files in a temporary directory on the fly (see arrow read/writer tests) for examples.

gawain-bolton replied:
Ok I have added unit tests for both the StreamReader and StreamWriter class.

emkornfield commented on cpp/examples/parquet/parquet-stream-api/stream-reader-writer.cc:
```diff
@@ -0,0 +1,213 @@
+#include "parquet/stream_reader.h"
+#include "parquet/stream_writer.h"
+
+class UserTimestamp {
```
document please.

gawain-bolton replied:
Ok

emkornfield commented on cpp/examples/parquet/parquet-stream-api/stream-reader-writer.cc:
```diff
@@ -0,0 +1,213 @@
+  return os;
+}
+
+std::shared_ptr<parquet::schema::GroupNode> get_schema() {
```
This doesn't conform with the style guide (as far as I know parquet like arrow is using the Google style guide).   This should be GetSchema()

gawain-bolton replied:
Ok

emkornfield commented on cpp/examples/parquet/parquet-stream-api/stream-reader-writer.cc:
```diff
@@ -0,0 +1,213 @@
+struct TestData {
+  static const int num_rows = 2000;
+
+  static std::string getString(const int i) { return "Str #" + std::to_string(i); }
```
style nit: GetString. (same for other method names in this class.

gawain-bolton replied:
Ok

emkornfield commented on cpp/examples/parquet/parquet-stream-api/stream-reader-writer.cc:
```diff
@@ -0,0 +1,213 @@
+char TestData::char4_array[] = "XYZ";
+std::time_t TestData::ts_offset;
+
+void write_parquet_file() {
```
styoe nit: WriteParquetFile

gawain-bolton replied:
Ok

emkornfield commented on cpp/examples/parquet/parquet-stream-api/stream-reader-writer.cc:
```diff
@@ -0,0 +1,213 @@
+    os << TestData::getUInt64(i);
+    os << TestData::getDouble(i);
+    os << TestData::getUserTimestamp(i);
+    os.nextRow();
```
it might be nice to provide an additional singleton class EndRow and EndRowGroup that can be inserted into the stream directly to make the api more consistent.

gawain-bolton replied:
Good idea!

emkornfield commented on cpp/examples/parquet/parquet-stream-api/stream-reader-writer.cc:
```diff
@@ -0,0 +1,213 @@
+  std::cout << "Parquet Stream Writing complete." << std::endl;
+}
+
+void read_parquet_file() {
```
ReadParquetFile

gawain-bolton replied:
Ok

emkornfield commented on cpp/examples/parquet/parquet-stream-api/stream-reader-writer.cc:
```diff
@@ -0,0 +1,213 @@
+  std::string s;
+  char ch;
+  char char_array[4];
+  std::int8_t int8;
```
generally we don't prefix primitive integer types with std::

emkornfield commented on cpp/src/parquet/stream_reader.cc:
```diff
@@ -0,0 +1,224 @@
+}
+
+StreamReader& StreamReader::operator>>(bool& v) {
+  checkColumn_(Type::BOOLEAN, ConvertedType::NONE);
```
style: CheckColumn

gawain-bolton replied:
Ok

emkornfield commented on cpp/src/parquet/stream_reader.cc:
```diff
@@ -0,0 +1,224 @@
+
+StreamReader& StreamReader::operator>>(bool& v) {
+  checkColumn_(Type::BOOLEAN, ConvertedType::NONE);
+  read_<BoolReader>(v);
```
style: just Read()

gawain-bolton replied:
Ok

emkornfield commented on cpp/src/parquet/stream_reader.cc:
```diff
@@ -0,0 +1,224 @@
+  auto* const schema{file_metadata_->schema()};
+  auto* const group_node{schema->group_node()};
+
+  nodes_.resize(schema->num_columns());
```
I think it is slightly more idiomatic to use reserve and push_back below (avoid unnecessary double initialization)

gawain-bolton replied:
I do not think the default constructor of a shared_ptr is expensive.
I prefer the code to make it absolutely clear that the nth element in the vector references the nth column.

emkornfield replied:
OK

emkornfield commented on cpp/src/parquet/stream_reader.cc:
```diff
@@ -0,0 +1,224 @@
+
+  nodes_.resize(schema->num_columns());
+
+  for (auto i = 0; i < schema->num_columns(); ++i) {
```
int64_t?

gawain-bolton replied:
The SchemaDescriptor::num_columns() function returns an integer so I do not think that is necessary.

xhochy replied:
Please try to avoid `auto` where possible. We strive to have the type of the variable at least once in the line it gets defined (for readability reasons).

emkornfield commented on cpp/src/parquet/stream_reader.cc:
```diff
@@ -0,0 +1,224 @@
+  std::memcpy(ptr, flba.ptr, len);
+}
+
+void StreamReader::read_(ByteArray& v) {
```
style: Read(ByteArray* )

gawain-bolton replied:
Ok

emkornfield commented on cpp/src/parquet/stream_reader.cc:
```diff
@@ -0,0 +1,224 @@
+  }
+}
+
+void StreamReader::read_(FixedLenByteArray& v) {
```
use pointer for output parameters when it isn't required for the API https://google.github.io/styleguide/cppguide.html#Output_Parameters

emkornfield commented on cpp/src/parquet/stream_reader.cc:
```diff
@@ -0,0 +1,224 @@
+                           std::to_string(column_index_) + " is invalid for " +
+                           std::to_string(nodes_.size()) + " columns");
+  }
+  const auto& node{nodes_[column_index_]};
```
spell out type.

gawain-bolton replied:
I don't understand this comment which seems to be referring to the column index check.
There is no type associated with an invalid column index.

bkietz replied:
I believe the suggestion is:
```suggestion
  const node_ptr_type& node{nodes_[column_index_]};
```

gawain-bolton replied:
Thanks for the clarification.
However, the whole point of the 'auto' keyword is to avoid spelling out types and make code easier to refactor.

emkornfield replied:
Please read https://google.github.io/styleguide/cppguide.html#Type_deduction for the reasoning.  There are pros and cons, however we do try to follow the style guide to avoid debates like this.

gawain-bolton replied:
The "style guide" says:
"For local variables, you can use type deduction to make the code clearer by eliminating type information that is obvious or irrelevant, so that the reader can focus on the meaningful parts of the code"
So why spell out a type which is irrelevant?

emkornfield replied:
The example in the style, guide give is eliding types on an iterator, but then recommenends the components of the iterator to spelled out types.  In this case without following the code, I'm not really sure what what node is (its an extra level of indirection).   There is some judgement here, so when a parquet committer does a final review they can give guidance.

emkornfield commented on cpp/src/parquet/stream_writer.h:
```diff
@@ -0,0 +1,135 @@
+
+namespace parquet {
+
+// A class for writing Parquet files using an output stream type API.
```
I think doxygen commenting is preferred.

gawain-bolton replied:
I do not see any doxygen for the existing classes in Parquet.

wesm replied:
That the current codebase has problems shouldn't stop us from doing the right thing with new code =) Someone needs to make a sweep over the public API and make everything Doxygen but it is not done yet

emkornfield commented on cpp/src/parquet/stream_writer.h:
```diff
@@ -0,0 +1,135 @@
+// nextRow() function.
+//
+// A maximum row group size can be configured, the default size is 512
+// Mio.  Alternatively the row group size can be set to zero and the
```
Mio? Is the requirement either or?  is 512 standard in the code base, that seems on the small side.

gawain-bolton replied:
https://parquet.apache.org/documentation/latest/ recommends the row group size be between 512 MB and 1GB
As for "Mio" is a "mebioctet" which is commonly used in Europe.  Can change to the old "MB" units if that's preferred.

emkornfield commented on cpp/src/parquet/stream_writer.h:
```diff
@@ -0,0 +1,135 @@
+
+  StreamWriter& operator<<(const char* v);
+
+  StreamWriter& operator<<(const std::string& v);
```
it might pay to make an override for arrow::string_view as well, but this might make circular dependencies worse.

gawain-bolton replied:
Ok I will add arrow::util::string_view as an output type to the StreamWriter class.

emkornfield replied:
Actually I think it might not work with constness, sorry for bad suggestion

gawain-bolton replied:
I do not see any issue with constness.

emkornfield commented on cpp/src/parquet/stream_reader.cc:
```diff
@@ -0,0 +1,258 @@
+
+namespace parquet {
+
+StreamReader::StreamReader(std::unique_ptr<ParquetFileReader>& File_Reader)
```
this should either be a raw pointer or a non-reference (i.e. do a ownership transfer).  If a raw pointer it should be clarified in the documentation that ownership is not taken.  It might be nice to provide an option to do both for users.

gawain-bolton replied:
It would be dangerous to have a raw pointer.   Ownership is taken by having a unique_ptr parameter which is released.    Ownership makes sense given that the destructor of this class closes the file.
If ownership should not be taken, a shared_ptr could be passed.   In this case, the destructor would be changed not to close the file.

emkornfield commented on cpp/src/parquet/stream_reader.h:
```diff
@@ -0,0 +1,159 @@
+ public:
+  StreamReader() = default;
+
+  explicit StreamReader(std::unique_ptr<ParquetFileReader>& File_Reader);
```
I commented on the implementation, make this a raw pointer and clarify owernship isn't transferred, or remove the reference and have users do an explicit move.

gawain-bolton replied:
Ok reference removed so user must be explicit move.

emkornfield commented on cpp/src/parquet/stream_reader.cc:
```diff
@@ -0,0 +1,258 @@
+      static_cast<ByteArrayReader*>(column_readers_[column_index_++].get())};
+  int64_t values_read;
+
+  reader->ReadBatch(1, nullptr, nullptr, v, &values_read);
```
it would be nice to use named parameters or comment with /*param_name=*/ for all literals.

emkornfield commented on cpp/src/parquet/stream_writer.h:
```diff
@@ -0,0 +1,181 @@
+
+  StreamWriter& operator<<(const std::chrono::microseconds& v);
+
+  StreamWriter& operator<<(const std::chrono::nanoseconds& v);
```
I believe int96 nanos is deprecated it might be worth noting that here.

gawain-bolton replied:
Thanks for letting me know!

emkornfield commented on cpp/src/parquet/stream_reader.cc:
```diff
@@ -0,0 +1,258 @@
+}
+
+StreamReader& StreamReader::operator>>(std::chrono::nanoseconds& v) {
+  CheckColumn(Type::INT96, ConvertedType::NONE);
```
Int96 is deprecated: https://issues.apache.org/jira/browse/PARQUET-323

gawain-bolton replied:
Ok, will remove this.

emkornfield commented on cpp/src/parquet/stream_reader.h:
```diff
@@ -0,0 +1,159 @@
+//
+class PARQUET_EXPORT StreamReader {
+ public:
+  StreamReader() = default;
```
is the class valid with the default constructor?

emkornfield commented on cpp/src/parquet/stream_writer.h:
```diff
@@ -0,0 +1,181 @@
+//
+class PARQUET_EXPORT StreamWriter {
+ public:
+  StreamWriter() = default;
```
is the object in a valid state if constructed with a default constructor?

gawain-bolton replied:
Please see reply given for same comment on StreamReader class.

emkornfield commented on cpp/src/parquet/stream_writer.h:
```diff
@@ -0,0 +1,181 @@
+
+  ~StreamWriter();
+
+  static void SetDefaultMaxRowGroupSize(const int64_t max_size);
```
we generally don't use const on primitive types. (https://google.github.io/styleguide/cppguide.html#Use_of_const)

gawain-bolton replied:
Although a const parameter passed by value has no effect on the caller, it does guarantee the implementationi cannot modify the parameter given to it, which is nice to know and avoids errors.

emkornfield replied:
*shrug* style ... same comment as above, we can let parquet committer be a tie breaker.  FWIW, you can keep const out of the declarations and only put them on the definition where they add this value (https://abseil.io/tips/109)

wesm replied:
I would prefer to not use const for primitive types for consistency with the rest of the codebase. It's a detail that need not be present in the public API

gawain-bolton commented:
> The CI failures look legitimate, can you take a look at fixing them.

Yes I am actively working on fixing all of the CI failures.

gawain-bolton commented:
I believe I have fixed all the CI errors related to this PR.
The remaining failures seem unrelated but please let me know if I am mistaken.

emkornfield commented on cpp/src/parquet/stream_reader.cc:
```diff
@@ -0,0 +1,255 @@
+
+namespace parquet {
+
+StreamReader::StreamReader(std::unique_ptr<ParquetFileReader> File_Reader)
```
File_Reader -> file_reader

emkornfield commented on cpp/src/parquet/stream_reader.cc:
```diff
@@ -0,0 +1,255 @@
+                           std::to_string(column_index_) + " is invalid for " +
+                           std::to_string(nodes_.size()) + " columns");
+  }
+  const auto& node = nodes_[column_index_];
```
for commiter to way in.  Should auto be spelled out to type (see conversation on previous commits).

emkornfield commented:
LGTM, module style comments.  @xhochy do you have time to do a final  review so this can be committed?

wesm commented on cpp/src/parquet/stream_writer.h:
```diff
@@ -0,0 +1,185 @@
+  //      assigned afterwards.
+  StreamWriter() = default;
+
+  explicit StreamWriter(std::unique_ptr<ParquetFileWriter> File_Writer);
```
non-conforming name

gawain-bolton replied:
Ok

wesm commented on cpp/src/parquet/stream_writer.h:
```diff
@@ -0,0 +1,185 @@
+  StreamWriter(const StreamWriter&) = delete;
+  StreamWriter& operator=(const StreamWriter&) = delete;
+
+  StreamWriter& operator<<(const bool v);
```
Ideally `const` would be removed from the various primitive types in this header

gawain-bolton replied:
Ok

wesm commented on cpp/src/parquet/stream_writer.h:
```diff
@@ -0,0 +1,185 @@
+
+  // Copying is not allowed.
+  StreamWriter(const StreamWriter&) = delete;
+  StreamWriter& operator=(const StreamWriter&) = delete;
```
We have a macro PARQUET_DISALLOW_COPY_AND_ASSIGN that you can use

gawain-bolton replied:
Unless this is mandatory I think that code is more clear without macros.

emkornfield replied:
IMO with C++11 it is cleaner to use "= delete" instead of the macro.

wesm commented on cpp/src/parquet/stream_writer.h:
```diff
@@ -0,0 +1,185 @@
+  StreamWriter& operator<<(const char v);
+
+  // Helper class to write fixed length strings.
+  struct PARQUET_EXPORT FixedStringView {
```
It's curious that this is required at all, could you comment why this helper class is needed when we have `arrow::util::string_view`?

gawain-bolton replied:
Yes and that class is used to write variable length string data.
However, another type is required to write fixed length data and so I think this helper class will be useful to distinguish this case.

wesm commented on cpp/src/parquet/stream_writer.h:
```diff
@@ -0,0 +1,185 @@
+#include <vector>
+
+#include "arrow/util/string_view.h"
+#include "parquet/api/writer.h"
```
Better to explicitly list the headers you depend on

gawain-bolton replied:
Ok

gawain-bolton commented on cpp/src/parquet/stream_reader.h:
```diff
@@ -0,0 +1,161 @@
+  // N.B. Default constructed objects are not usable.  This
+  //      constructor is provided so that the object may be move
+  //      assigned afterwards.
+  StreamReader() = default;
```
Yes this is useful to avoid having to use pointers and do dynamic memory allocation.

gawain-bolton commented on cpp/src/parquet/stream_reader.h:
```diff
@@ -0,0 +1,161 @@
+
+namespace parquet {
+
+// A class for reading Parquet files using an output stream type API.
```
Ok

gawain-bolton commented on cpp/src/parquet/stream_reader.h:
```diff
@@ -0,0 +1,161 @@
+// The user must explicitly advance to the next row using the
+// EndRow() function or EndRow input manipulator.
+//
+// Currently there is no support for optional or repeated fields.
```
I understand and I would like to add support for optional fields in the future.
I think support for optional fields could easily be added by using std::optional (surprisingly enough!), however that requires C++17 and from what I see arrow is using C++11.

gawain-bolton commented on cpp/src/parquet/stream_reader.h:
```diff
@@ -0,0 +1,161 @@
+#include <string>
+#include <vector>
+
+#include "parquet/api/reader.h"
```
Ok

gawain-bolton commented on cpp/src/parquet/stream_reader.cc:
```diff
@@ -0,0 +1,255 @@
+    : file_reader_{std::move(File_Reader)} {
+  file_metadata_ = file_reader_->metadata();
+
+  auto* const schema = file_metadata_->schema();
```
Ok

gawain-bolton commented:
> +1. Thanks for your patience and for being diligent about addressing the comments, really appreciate this contribution!

Thanks to you for allowing me to participate in this very useful and interesting project!
I look forward to contributing more in the future!

kou commented on cpp/src/parquet/stream_writer.h:
```diff
@@ -0,0 +1,187 @@
+
+  // Moving is possible.
+  StreamWriter(StreamWriter&&) noexcept = default;
+  StreamWriter& operator=(StreamWriter&&) noexcept = default;
```
This causes build failure with g++ 4.8.5 on CentOS 7: https://issues.apache.org/jira/browse/PARQUET-1688

It seems that g++ 4.8.5 doesn't have the default implementation of `operator=() noexcept`:

```text
In file included from /root/rpmbuild/BUILD/apache-arrow-1.0.0/cpp/src/parquet/stream_reader.h:31:0,
                 from /root/rpmbuild/BUILD/apache-arrow-1.0.0/cpp/src/parquet/stream_reader.cc:18:
/root/rpmbuild/BUILD/apache-arrow-1.0.0/cpp/src/parquet/stream_writer.h:67:17: error: function 'parquet::StreamWriter& parquet::StreamWriter::operator=(parquet::StreamWriter&&)' defaulted on its first declaration with an exception-specification that differs from the implicit declaration 'parquet::StreamWriter& parquet::StreamWriter::operator=(parquet::StreamWriter&&)'
   StreamWriter& operator=(StreamWriter&&) noexcept = default;
                 ^
In file included from /root/rpmbuild/BUILD/apache-arrow-1.0.0/cpp/src/parquet/stream_reader.cc:18:0:
/root/rpmbuild/BUILD/apache-arrow-1.0.0/cpp/src/parquet/stream_reader.h:61:17: error: function 'parquet::StreamReader& parquet::StreamReader::operator=(parquet::StreamReader&&)' defaulted on its first declaration with an exception-specification that differs from the implicit declaration 'parquet::StreamReader& parquet::StreamReader::operator=(parquet::StreamReader&&)'
   StreamReader& operator=(StreamReader&&) noexcept = default;
                 ^
make[2]: *** [src/parquet/CMakeFiles/parquet_objlib.dir/stream_reader.cc.o] Error 1
make[2]: *** Waiting for unfinished jobs....
In file included from /root/rpmbuild/BUILD/apache-arrow-1.0.0/cpp/src/parquet/stream_writer.cc:18:0:
/root/rpmbuild/BUILD/apache-arrow-1.0.0/cpp/src/parquet/stream_writer.h:67:17: error: function 'parquet::StreamWriter& parquet::StreamWriter::operator=(parquet::StreamWriter&&)' defaulted on its first declaration with an exception-specification that differs from the implicit declaration 'parquet::StreamWriter& parquet::StreamWriter::operator=(parquet::StreamWriter&&)'
   StreamWriter& operator=(StreamWriter&&) noexcept = default;
                 ^
```
